clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -491,30 +494,65 @@ protected void processMessageAndMaybeProduceToKafka(
       // call in this context much less obtrusive, however, it implies that all views can only work for AA stores
       int valueSchemaId =
           rmdWithValueSchemaID != null ? rmdWithValueSchemaID.getValueSchemaId() : incomingValueSchemaId;
-      this.viewWriters.forEach(
-          (k, v) -> v.processRecord(
-              mergeConflictResult.getNewValue(),
-              oldValueProvider.get(),
-              keyBytes,
-              versionNumber,
-              incomingValueSchemaId,
-              valueSchemaId,
-              mergeConflictResult.getRmdRecord()));
 
-      // This function may modify the original record in KME and it is unsafe to use the payload from KME directly after
-      // this call.
-      producePutOrDeleteToKafka(
-          mergeConflictResult,
-          partitionConsumptionState,
-          keyBytes,
-          consumerRecord,
-          subPartition,
-          kafkaUrl,
-          kafkaClusterId,
-          beforeProcessingRecordTimestampNs,
-          valueManifestContainer.getManifest(),
-          rmdWithValueSchemaID == null ? null : rmdWithValueSchemaID.getRmdManifest());
+      // Get Futures and
+      // Pass callback with on completion which decrements, once all decrement call producePutOrDeleteToKafka
+      if (this.viewWriters.size() > 0) {
+        Long preprocessingTime = System.currentTimeMillis();
+
+        List<Future> viewWriteCallbacks = new ArrayList<>();
+        for (VeniceViewWriter writer: viewWriters.values()) {
+          viewWriteCallbacks.add(
+              writer.processRecord(
+                  mergeConflictResult.getNewValue(),
+                  oldValueProvider.get(),
+                  keyBytes,
+                  versionNumber,
+                  incomingValueSchemaId,
+                  valueSchemaId,
+                  mergeConflictResult.getRmdRecord()));
+        }
+        viewWriteCallbacks.add(partitionConsumptionState.getLastVTProduceCallFuture());
+        CompletableFuture.allOf(viewWriteCallbacks.toArray(new CompletableFuture[0]))
```

Comment:
Since `CompletableFuture.allOf` expects an array, can we just create that direct directly, instead of passing by an `ArrayList` first? We can initialize the array at the size of the views map + 1. That would save one object and one array copy from each write...